### PR TITLE
Fix group expression not replace lowestCostExpressions in the group when group merge

### LINKED ISSUE #4917: Query returns Unknown error
### Steps to reproduce the behavior (Required)

1. CREATE EXTERNAL RESOURCE "hive_resource_emr" PROPERTIES ("type"="hive", "hive.metastore.uris"="thrift://172.26.194.238:9083")
2. CREATE EXTERNAL RESOURCE "hive2" PROPERTIES ("type"="hive","hive.metastore.uris"="thrift://172.26.92.141:9083")
3. CREATE TABLE `tpcds_100g_call_center` (
  `cc_call_center_sk` int(11) NOT NULL COMMENT "",
  `cc_call_center_id` char(16) NOT NULL COMMENT "",
  `cc_rec_start_date` date NULL COMMENT "",
  `cc_rec_end_date` date NULL COMMENT "",
  `cc_closed_date_sk` int(11) NULL COMMENT "",
  `cc_open_date_sk` int(11) NULL COMMENT "",
  `cc_name` varchar(50) NULL COMMENT "",
  `cc_class` varchar(50) NULL COMMENT "",
  `cc_employees` int(11) NULL COMMENT "",
  `cc_sq_ft` int(11) NULL COMMENT "",
  `cc_hours` char(20) NULL COMMENT "",
  `cc_manager` varchar(40) NULL COMMENT "",
  `cc_mkt_id` int(11) NULL COMMENT "",
  `cc_mkt_class` char(50) NULL COMMENT "",
  `cc_mkt_desc` varchar(100) NULL COMMENT "",
  `cc_market_manager` varchar(40) NULL COMMENT "",
  `cc_division` int(11) NULL COMMENT "",
  `cc_division_name` varchar(50) NULL COMMENT "",
  `cc_company` int(11) NULL COMMENT "",
  `cc_company_name` char(50) NULL COMMENT "",
  `cc_street_number` char(10) NULL COMMENT "",
  `cc_street_name` varchar(60) NULL COMMENT "",
  `cc_street_type` char(15) NULL COMMENT "",
  `cc_suite_number` char(10) NULL COMMENT "",
  `cc_city` varchar(60) NULL COMMENT "",
  `cc_county` varchar(30) NULL COMMENT "",
  `cc_state` char(2) NULL COMMENT "",
  `cc_zip` char(10) NULL COMMENT "",
  `cc_country` varchar(20) NULL COMMENT "",
  `cc_gmt_offset` decimal64(5, 2) NULL COMMENT "",
  `cc_tax_percentage` decimal64(5, 2) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`cc_call_center_sk`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`cc_call_center_sk`) BUCKETS 5
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
4. CREATE EXTERNAL TABLE `hive_ssb_100g_orc_customer` (
  `c_custkey` int(11) NULL COMMENT "",
  `c_name` varchar(26) NULL COMMENT "",
  `c_address` varchar(41) NULL COMMENT "",
  `c_city` varchar(11) NULL COMMENT "",
  `c_nation` varchar(16) NULL COMMENT "",
  `c_region` varchar(13) NULL COMMENT "",
  `c_phone` varchar(16) NULL COMMENT "",
  `c_mktsegment` varchar(11) NULL COMMENT ""
) ENGINE=HIVE
COMMENT "PARTITION BY ()"
PROPERTIES (
"database" = "ssb_orc",
"table" = "customer",
"resource" = "hive_resource_emr",
"hive.metastore.uris"  =  "thrift://172.26.194.238:9083"
);
5. CREATE EXTERNAL TABLE `hive_ssb_100g_csv_part` (
  `p_partkey` int(11) NULL COMMENT "",
  `p_name` varchar(23) NULL COMMENT "",
  `p_mfgr` varchar(7) NULL COMMENT "",
  `p_category` varchar(8) NULL COMMENT "",
  `p_brand` varchar(10) NULL COMMENT "",
  `p_color` varchar(12) NULL COMMENT "",
  `p_type` varchar(26) NULL COMMENT "",
  `p_size` int(11) NULL COMMENT "",
  `p_container` varchar(11) NULL COMMENT ""
) ENGINE=HIVE
COMMENT "PARTITION BY ()"
PROPERTIES (
"database" = "ssb_100g_csv",
"table" = "part",
"resource" = "hive_resource_emr",
"hive.metastore.uris"  =  "thrift://172.26.194.238:9083"
);
6. CREATE EXTERNAL TABLE `hive_external_table` (
  `col_date` date NULL COMMENT "column date",
  `col_tinyint` tinyint(4) NULL COMMENT "column tinyint",
  `col_string` varchar(1000) NULL COMMENT "column string",
  `col_smallint` smallint(6) NULL COMMENT "column smallint",
  `col_int` int(11) NULL COMMENT "column int",
  `col_integer` int(11) NULL COMMENT "column integer",
  `col_bigint` bigint(20) NULL COMMENT "column bigint",
  `col_float` float NULL COMMENT "column float",
  `col_double` double NULL COMMENT "column double",
  `col_double_precision` double NULL COMMENT "column double precision",
  `col_decimal` decimal64(10, 0) NULL COMMENT "column decimal",
  `col_timestamp` datetime NULL COMMENT "column timestamp",
  `col_varchar` varchar(100) NULL COMMENT "colu
mn varchar",
  `col_binary` varchar(1000) NULL COMMENT "column binary",
  `col_boolean` boolean NULL COMMENT "column boolean"
) ENGINE=HIVE
COMMENT "PARTITION BY ()"
PROPERTIES (
"database" = "hive_extbl_test_1636019895543",
"table" = "hive_hdfs_orc_nocompress",
"resource" = "hive_resource",
"hive.metastore.uris"  =  "thrift://172.26.92.141:9083"
);
7. CREATE TABLE `tpch_100g_customer` (
  `C_CUSTKEY` int(11) NOT NULL COMMENT "",
  `C_NAME` varchar(25) NOT NULL COMMENT "",
  `C_ADDRESS` varchar(40) NOT NULL COMMENT "",
  `C_NATIONKEY` int(11) NOT NULL COMMENT "",
  `C_PHONE` char(15) NOT NULL COMMENT "",
  `C_ACCTBAL` decimal64(15, 2) NOT NULL COMMENT "",
  `C_MKTSEGMENT` char(10) NOT NULL COMMENT "",
  `C_COMMENT` varchar(117) NOT NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`C_CUSTKEY`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`C_CUSTKEY`) BUCKETS 12
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "DEFAULT"
);
8. CREATE EXTERNAL TABLE `hive_ssb_1g_csv_lineorder` (
  `lo_orderkey` bigint(20) NULL COMMENT "",
  `lo_linenumber` int(11) NULL COMMENT "",
  `lo_custkey` int(11) NULL COMMENT "",
  `lo_partkey` int(11) NULL COMMENT "",
  `lo_suppkey` int(11) NULL COMMENT "",
  `lo_orderdate` int(11) NULL COMMENT "",
  `lo_orderpriority` varchar(16) NULL COMMENT "",
  `lo_shippriority` int(11) NULL COMMENT "",
  `lo_quantity` int(11) NULL COMMENT "",
  `lo_extendedprice` int(11) NULL COMMENT "",
  `lo_ordtotalprice` int(11) NULL COMMENT "",
  `lo_discount` int(11) NULL COMMENT "",
  `lo_revenue` int(11) NULL COMMENT "",
  `lo_supplycost` int(11) NULL COMMENT "",
  `lo_tax` int(11) NULL COMMENT "",
  `lo_commitdate` int(11) NULL COMMENT "",
  `lo_shipmode` varchar(11) NULL COMMENT ""
) ENGINE=HIVE
COMMENT "PARTITION BY ()"
PROPERTIES (
"database" = "ssb_1g_csv",
"table" = "lineorder",
"resource" = "hive2",
"hive.metastore.uris"  =  "thrift://172.26.92.141:9083"
);
9. select   * from   hive_ssb_100g_orc_customer as ref_0   right join tpcds_100g_call_center as ref_1 on (ref_0.c_name = ref_1.cc_call_center_id)   inner join hive_ssb_100g_csv_part as ref_2 on (ref_0.c_region = ref_2.p_name)   inner join hive_external_table as ref_3 on (ref_2.p_mfgr = ref_3.col_string)   inner join tpch_100g_customer as ref_4 on (ref_3.col_string = ref_4.C_NAME)   inner join hive_ssb_1g_csv_lineorder as ref_5 on (ref_1.cc_call_center_sk = ref_5.lo_orderkey) where   ref_5.lo_revenue <> ref_0.c_custkey;

### Expected behavior (Required)
Query returns correctly

### Real behavior (Required)
```
2022-04-08 15:58:42,377 WARN (starrocks-mysql-nio-pool-6|462) [StmtExecutor.execute():473] execute Exception, sql select   * from   sqlsmith.hive_ssb_100g_orc_customer as ref_0   right join sqlsmith.tpcds_100g_call_center as ref_1 on (ref_0.c_name = ref_1.cc_call_center_id)   inner join sqlsmith.hive_ssb_100g_csv_part as ref_2 on (ref_0.c_region = ref_2.p_name)   inner join sqlsmith.hive_external_table as ref_3 on (ref_2.p_mfgr = ref_3.col_string)   inner join sqlsmith.tpch_100g_customer as ref_4 on (ref_3.col_string = ref_4.C_NAME)   inner join sqlsmith.hive_ssb_1g_csv_lineorder as ref_5 on (ref_1.cc_call_center_sk = ref_5.lo_orderkey) where   ref_5.lo_revenue <> ref_0.c_custkey
java.lang.StackOverflowError: null
	at java.util.Arrays.hashCode(Arrays.java:4146) ~[?:1.8.0_292]
	at java.util.Objects.hash(Objects.java:128) ~[?:1.8.0_292]
	at com.starrocks.sql.optimizer.base.PhysicalPropertySet.hashCode(PhysicalPropertySet.java:64) ~[starrocks-fe.jar:?]
	at java.util.HashMap.hash(HashMap.java:339) ~[?:1.8.0_292]
	at java.util.HashMap.containsKey(HashMap.java:596) ~[?:1.8.0_292]
	at com.starrocks.sql.optimizer.Group.hasBestExpression(Group.java:174) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Group.getBestExpression(Group.java:167) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:259) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.extractBestPlan(Optimizer.java:264) ~[starrocks-fe.jar:?]
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
branch-2.2 release 46db7d9

### PR DESCRIPTION
**Pull request by Youngwb**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：

Fixes #4917 

## Problem Summary(Required) ：

When Group merge other goup,  the needReinsertedExpressions should consider the replacement of lowestCostExpressions in the Group, otherwise the optimizer will use wrong groupExpression when extractBestPlan

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|16|16|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|3|3|100.00%|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java [hunk 1]
```diff
@@ -151,8 +151,6 @@ private boolean needMerge(Group targetGroup, Group existedGroup) {
 
     // Merge srcGroup to dstGroup, srcGroup will be deleted
     private void mergeGroup(Group srcGroup, Group dstGroup) {
-        groups.remove(srcGroup);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:154]: why not remove src group?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java [hunk 2]
```diff
@@ -198,11 +196,16 @@ private void mergeGroup(Group srcGroup, Group dstGroup) {
                 // set unused.
                 groupExpression.getGroup().removeGroupExpression(groupExpression);
                 groupExpression.setUnused(true);
+                GroupExpression existGroupExpression = groupExpressions.get(groupExpression);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:199]: If the hash map only compute hash value once, or we cache the hash value, the merge process seems should be more simple.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java (hunk 1 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:154]: I clean the group which will be merged, it will be clean by empty group

Comment (kangkaisen) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:154]: removeOneGroup method is more complex and unnecessary. I don't think this change is better.

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:154]: could we remove the removeOneGroup method? if can not, this way is better

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java (hunk 2 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Memo.java:199]: It is not more simple, if we cache the hash value,  at merge group phase, we also need to consider the group expresions which have same hashcode

**Review by Seaven — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 19 / 19 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Group.java|16|16|100.00%|
|:large_blue_circle:|com/starrocks/sql/optimizer/Memo.java|3|3|100.00%|

**Review by Seaven — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Comment by Youngwb** [comment]
@Mergifyio backport branch-2.2

**Pull request merged**